subprojects/core/src/main/java/org/gradle/api/internal/MutationGuards.java

```diff
@@ -77,4 +80,23 @@ public static MutationGuard of(Object target) {
             return IDENTITY_MUTATION_GUARD;
         }
     }
+
+    /**
+     * Retrieves the {@code MutationGuard} of the targets if it implements {@code WithMutationGuard}, else returns an identity mutation guard performing no guard operations.
+     */
+    public static MutationGuard of(Iterable<Object> targets) {
```

Comment:
Why multiple targets? This look like a lot of complexity (and performance cost). I'm not sure why we would need more than one target at any given time.

Reply:
All related to `CompositeDomainObjectSet`. It was used so external code can disable the guard if they know what they are doing. I understand that if the fix to `BasePlugin` works appropriately, we don't really need this feature. I will revert this change.